### PR TITLE
skip traversing raw blocks while garbage collecting

### PR DESCRIPTION
**Pull request by balajijinnah**
Fix for #7160

Signed-off-by: Tiger <rbalajis25@gmail.com>

### CODE REVIEW

File: gc/gc.go [hunk 1]
```diff
@@ -172,17 +172,26 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
 	// disk backed to conserve memory.
 	errors := false
 	gcs := cid.NewSet()
-	getLinks := func(ctx context.Context, cid cid.Cid) ([]*ipld.Link, error) {
-		links, err := ipld.GetLinks(ctx, ng, cid)
+	getLinks := func(ctx context.Context, CID cid.Cid) ([]*ipld.Link, error) {
```
Comment (Stebalien) [inline gc/gc.go:175]: Please revert the variable name change.

File: gc/gc.go [hunk 2]
```diff
@@ -172,17 +172,26 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
+
+		filteredLinks := []*ipld.Link{}
+		for _, link := range links {
+			// Skip raw links, since it doesn't contain any leaf links to traverse.
+			if link.Cid.Type() == cid.Raw {
+				continue
+			}
+			filteredLinks = append(filteredLinks, link)
+		}
+		return filteredLinks, nil
```
Comment (Stebalien) [inline gc/gc.go:194]: A simpler (and more reliable) approach would be to filter CIDs of raw blocks at the _top_ of this function. That is, a raw `cid` is passed in, return `nil, nil` (no links, no error).

File: gc/gc.go [hunk 3]
```diff
@@ -198,17 +207,26 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
```
Comment (Stebalien) [inline gc/gc.go:229]: ditto about moving this check to the top.

File: gc/gc.go (hunk 1 above)
Comment (balajijinnah) [reply gc/gc.go:175]: If I use cid as variable, variable is shadowing the cid package. 
I'm not able to use cid.Raw

Comment (Stebalien) [reply gc/gc.go:175]: Ah, then I'd use `id, `c` or `key`.  Upper-case is usually reserved for exported variables.

File: gc/gc.go [hunk 4]
```diff
@@ -172,12 +172,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
 	// disk backed to conserve memory.
 	errors := false
 	gcs := cid.NewSet()
-	getLinks := func(ctx context.Context, cid cid.Cid) ([]*ipld.Link, error) {
-		links, err := ipld.GetLinks(ctx, ng, cid)
+	getLinks := func(ctx context.Context, CID cid.Cid) ([]*ipld.Link, error) {
```
Comment (Stebalien) [inline gc/gc.go:175]:
```suggestion
	getLinks := func(ctx context.Context, link cid.Cid) ([]*ipld.Link, error) {
```

File: gc/gc.go [hunk 5]
```diff
@@ -172,12 +172,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
+		if CID.Type() == cid.Raw {
```
Comment (Stebalien) [inline gc/gc.go:176]:
```suggestion
		if link.Type() == cid.Raw {
```

File: gc/gc.go [hunk 6]
```diff
@@ -172,12 +172,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
+			return nil, nil
+		}
+
+		links, err := ipld.GetLinks(ctx, ng, CID)
```
Comment (Stebalien) [inline gc/gc.go:180]:
```suggestion
		links, err := ipld.GetLinks(ctx, ng, link)
```

File: gc/gc.go [hunk 7]
```diff
@@ -172,12 +172,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
 		if err != nil {
 			errors = true
 			select {
-			case output <- Result{Error: &CannotFetchLinksError{cid, err}}:
+			case output <- Result{Error: &CannotFetchLinksError{CID, err}}:
```
Comment (Stebalien) [inline gc/gc.go:184]:
```suggestion
			case output <- Result{Error: &CannotFetchLinksError{link, err}}:
```

File: gc/gc.go [hunk 8]
```diff
@@ -198,12 +202,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
 		}
 	}
 
-	bestEffortGetLinks := func(ctx context.Context, cid cid.Cid) ([]*ipld.Link, error) {
-		links, err := ipld.GetLinks(ctx, ng, cid)
+	bestEffortGetLinks := func(ctx context.Context, CID cid.Cid) ([]*ipld.Link, error) {
```
Comment (Stebalien) [inline gc/gc.go:205]:
```suggestion
	bestEffortGetLinks := func(ctx context.Context, link cid.Cid) ([]*ipld.Link, error) {
```

File: gc/gc.go [hunk 9]
```diff
@@ -198,12 +202,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
+		if CID.Type() == cid.Raw {
```
Comment (Stebalien) [inline gc/gc.go:206]:
```suggestion
		if link.Type() == cid.Raw {
```

File: gc/gc.go [hunk 10]
```diff
@@ -198,12 +202,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
+			return nil, nil
+		}
+
+		links, err := ipld.GetLinks(ctx, ng, CID)
```
Comment (Stebalien) [inline gc/gc.go:210]:
```suggestion
		links, err := ipld.GetLinks(ctx, ng, link)
```

File: gc/gc.go [hunk 11]
```diff
@@ -198,12 +202,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
 		if err != nil && err != ipld.ErrNotFound {
 			errors = true
 			select {
-			case output <- Result{Error: &CannotFetchLinksError{cid, err}}:
+			case output <- Result{Error: &CannotFetchLinksError{CID, err}}:
```
Comment (Stebalien) [inline gc/gc.go:214]:
```suggestion
			case output <- Result{Error: &CannotFetchLinksError{link, err}}:
```

File: gc/gc.go [hunk 12]
```diff
@@ -172,12 +172,16 @@ func ColoredSet(ctx context.Context, pn pin.Pinner, ng ipld.NodeGetter, bestEffo
...
 	errors := false
 	gcs := cid.NewSet()
-	getLinks := func(ctx context.Context, cid cid.Cid) ([]*ipld.Link, error) {
-		links, err := ipld.GetLinks(ctx, ng, cid)
+	getLinks := func(ctx context.Context, link cid.Cid) ([]*ipld.Link, error) {
+		if link.Type() == cid.Raw {
+			return nil, nil
+		}
+
+		links, err := ipld.GetLinks(ctx, ng, link)
```
Comment (sinkuu) [inline gc/gc.go:180]: `GetLinks` seems to already have fast path for raw blocks.

```go
func GetLinks(ctx context.Context, ng NodeGetter, c cid.Cid) ([]*Link, error) {
	if c.Type() == cid.Raw {
		return nil, nil
	}
	...
```

Comment (Stebalien) [reply gc/gc.go:180]: Well... Now really I feel like an idiot.

@balajijinnah I'm so sorry. Apparently, we already _have_ this optimization, I just never thought to actually check.

Comment (balajijinnah) [reply gc/gc.go:180]: Lol, Even I didn't even check it. 

It's a good experience to contribute. Looking to contribute more.

**Comment by Stebalien** [comment]
Well, I'm not going to say no to that! Depending on how familiar you are
with go, you might want to try
https://github.com/ipfs/go-ipfs/issues/7227 or https://github.com/ipfs/go-ipfs/issues/7269.

**Pull request closed without merging**